### ISSUE TITLE
Unable to register multiple serializers with same response wrapper class

### ISSUE THREAD

**Issue by AndWa** [ktorio/ktor#1019]
### Feedback

I am playing around with Kotlin Native multiplatform and testing it by implementing a simple Android and iOS version of an app.

I am using ktor for my network requests and it works great but I experience issues when I try to register multiple serializers that have the same response wrapper class.

Example.

```json
{
    "data": {
        "firstName": "John",
        "lastName": "Doe",
        "age": 28
    }
}
```

Kotlin data class representation of the json

```kotlin
@Serializable
data class ResponseWrapper<T>(val data: T?)

@Serializable
data class Person (
    val firstName: String,
    val lastName: String,
    val age: Int
)
```

I am using [Kodein](https://github.com/Kodein-Framework/Kodein-DI) for dependency injection and I am binding my ```HttpClient``` like this

```kotlin
 bind<HttpClient>() with provider {
        HttpClient {
            install(JsonFeature) {
                serializer = KotlinxSerializer(Json.nonstrict).apply {
                    register(ResponseWrapper.serializer(Person.serializer()))
                }
            }
            install(ExpectSuccess)
        }
    }
```

My register basically registers ```ResponseWrapper<Person>``` and it **works great**.

But I encounter a problem once I try to add another Serializable class.

Example.

```json
{
    "data": {
        "name": "Somewhere",
        "temperature": 17.5
    }
}
```

and Kotlin data class representation of the json

```kotlin
@Serializable
data class City (
    val name: String,
    val temperature: Double
)
```

I register this class as well

```kotlin
 bind<HttpClient>() with provider {
        HttpClient {
            install(JsonFeature) {
                serializer = KotlinxSerializer(Json.nonstrict).apply {
                    register(ResponseWrapper.serializer(Person.serializer()))
                    register(ResponseWrapper.serializer(City.serializer()))
                }
            }
            install(ExpectSuccess)
        }
    }
```

If I try to do a request now only the last registered serializer will work ```                register(ResponseWrapper.serializer(City.serializer()))```

As much as I was able to figure out this issue is happening because of https://github.com/ktorio/ktor/blob/4cf2d38bf7de5357d239a9480bbe83ba6250e86f/ktor-client/ktor-client-features/ktor-client-json/common/src/io/ktor/client/features/json/serializer/KotlinxSerializer.kt#L36

Since the key that will be used is the type, in case of ```ResponseWrapper<Person>``` and ```ResponseWrapper<City>``` for both of these cases the key ends up being the ```ResponseWrapper``` class and because of that only the last serializer is saved into the ```mappers``` since it overwrites existing value.

Is this something that should be supported or am I doing something wrong here and is there a better way to do this?

**Comment by AndWa** [comment]
@e5l do you have an estimate when will this be fixed and is there any workaround in the meanwhile?

**Comment by e5l** [comment]
Hi @AndWa, sorry but the fix blocked by https://youtrack.jetbrains.com/issue/KT-28230

**Comment by kuuuurt** [comment]
@e5l any news on this? It's still not working as of 1.2.5

**Comment by e5l** [comment]
Fixed with `typeOf` support in 1.3.0
